File: tests/identity/fakes/fake_person_repository.py
from trocado.features.identity.application.interfaces.person_repository_interface import (
    PersonRepositoryInterface,
)
from trocado.features.identity.domain.entities.person_entity import PersonEntity
from trocado.features.identity.domain.enums.person_status import PersonStatus
from trocado.features.identity.domain.value_objects.email_value_object import EmailValueObject


class FakePersonRepository(PersonRepositoryInterface):
    """In-memory test double. Stores people in a list; reads see only active accounts."""

    def __init__(self, *people: PersonEntity) -> None:
        self.people: list[PersonEntity] = list(people)

    async def find_active_by_id(self, person_id: str) -> PersonEntity | None:
        return next(
            (person for person in self.people if person.status is PersonStatus.ACTIVE and person.id == person_id),
            None,
        )

    async def find_active_by_email(self, email: EmailValueObject) -> PersonEntity | None:
        return next(
            (person for person in self.people if person.status is PersonStatus.ACTIVE and person.email == email),
            None,
        )

    async def create(self, person: PersonEntity) -> None:
        self.people.append(person)

    async def update(self, person: PersonEntity) -> None:
        self.people = [person if existing.id == person.id else existing for existing in self.people]

    async def delete(self, person: PersonEntity) -> None:
        self.people = [person if existing.id == person.id else existing for existing in self.people]

File: src/test/kotlin/com/bed/cordato/features/identity/factories/FakePersonRepository.kt
package com.bed.cordato.features.identity.factories

import com.bed.cordato.features.identity.domain.entities.PersonEntity
import com.bed.cordato.features.identity.domain.enums.PersonStatusEnum
import com.bed.cordato.features.identity.domain.value_objects.NameValueObject
import com.bed.cordato.features.identity.domain.value_objects.EmailValueObject

import com.bed.cordato.features.identity.application.driven.outcomes.UpdateEmailOutcome
import com.bed.cordato.features.identity.application.driven.repositories.PersonRepository

class FakePersonRepository : PersonRepository {
    private val byEmail = mutableMapOf<String, PersonEntity>()

    override fun existsByEmail(email: EmailValueObject): Boolean = byEmail.containsKey(email.value)

    override fun signUp(person: PersonEntity): Boolean =
        byEmail.putIfAbsent(person.email.value, person) == null

    override fun findByEmail(email: EmailValueObject): PersonEntity? =
        byEmail[email.value]?.takeIf { it.status == PersonStatusEnum.ACTIVE }

    override fun findById(id: String): PersonEntity? =
        byEmail.values.firstOrNull { it.id == id }?.takeIf { it.status == PersonStatusEnum.ACTIVE }

    override fun updateName(id: String, name: NameValueObject): Boolean {
        val person = findById(id) ?: return false
        byEmail[person.email.value] = person.copy(name = name)
        return true
    }

    override fun updateEmail(id: String, email: EmailValueObject): UpdateEmailOutcome {
        val person = findById(id) ?: return UpdateEmailOutcome.PERSON_INACTIVE
        val owner = byEmail[email.value]
        if (owner != null && owner.id != person.id) return UpdateEmailOutcome.EMAIL_TAKEN
        byEmail.remove(person.email.value)
        byEmail[email.value] = person.copy(email = email)
        return UpdateEmailOutcome.UPDATED
    }

    override fun updatePassword(id: String, hash: String): Boolean {
        val person = findById(id) ?: return false
        byEmail[person.email.value] = person.copy(hash = hash)
        return true
    }

    override fun deleteAccount(id: String): Boolean {
        val person = findById(id) ?: return false
        val neutralizedEmail = checkNotNull(EmailValueObject.of("deleted+$id@deleted.invalid"))

        byEmail.remove(person.email.value)
        byEmail[neutralizedEmail.value] = person.copy(email = neutralizedEmail, status = PersonStatusEnum.DELETED)
        return true
    }
}
